### PR TITLE
fix: Differentiate between Fulfilled and Completed

### PR DESCRIPTION
**Pull request by simster7**
Fixes: #3081

### CODE REVIEW

**Comment by simster7** [comment]
> tests!

Was already working on them when you made your review 🙂

**Comment by alexec** [comment]
The semantics of "Complete()" are wrong - how can something that was never started ever be completed? I think `Completed()` should exclude skipped nodes. 

Example:

Alex: "I have completed the washing up"
GF: "Alex - I can see that the sink is full of dishes - how can you claim you did the washing up?"
Alex: "Obviously, by 'complete' I meant that I skipped the washing up today"
GF: "Stop being silly - please complete the washing up"

How about:

rename `Completed()` to `CompletedOrSkipped()` 
rename `Executed()` to `Completed()`

**Comment by simster7** [comment]
> The semantics of "Complete()" are wrong - how can something that was never started ever be completed? I think Completed() should exclude skipped nodes.

Completely agreed. I had the same thought but wanted to avoid changing the name if necessary.

How about:

* Change `Completed()` to `Fulfilled()`. I.e., this node's purpose has been fulfilled, we don't have to consider it again.

* Change `Executed()` to `Completed()`

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -1189,16 +1189,21 @@ func (n Nodes) GetResourcesDuration() ResourcesDuration {
...
 func (phase NodePhase) Completed() bool {
 	return phase == NodeSucceeded ||
 		phase == NodeFailed ||
-		phase == NodeError ||
-		phase == NodeSkipped
+		phase == NodeError
 }
 
 // Completed returns whether or not the workflow has completed execution
 func (ws WorkflowStatus) Completed() bool {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1205]: don't you want to update this as well?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -1189,16 +1189,21 @@ func (n Nodes) GetResourcesDuration() ResourcesDuration {
 	return i
 }
 
+// Fulfilled returns whether a phase is fulfilled, i.e. it finished execution or was skipped
+func (phase NodePhase) Fulfilled() bool {
+	return phase.Completed() || phase == NodeSkipped
+}
+
+// Completed returns whether or not a phase completed. Notably, a skipped phase is not considered as having completed
 func (phase NodePhase) Completed() bool {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1198]: * [ ] todo - double check every instance is renamed

File: util/printer/workflow-printer.go [hunk 3]
```diff
@@ -89,7 +89,7 @@ func countPendingRunningCompleted(wf *wfv1.Workflow) (int, int, int) {
 		if tmpl == nil || !tmpl.IsPodType() {
 			continue
 		}
-		if node.Completed() {
+		if node.Fulfilled() {
 			completed++
```
Comment (alexec) [inline util/printer/workflow-printer.go:93]: rename variable?

File: workflow/controller/dag.go [hunk 4]
```diff
@@ -377,12 +377,14 @@ func (woc *wfOperationCtx) executeDAGTask(dagCtx *dagContext, taskName string) {
...
 	node := dagCtx.getTaskNode(taskName)
 	task := dagCtx.GetTask(taskName)
-	if node != nil && node.Completed() {
-		// Run the node's onExit node, if any.
-		hasOnExitNode, onExitNode, err := woc.runOnExitNode(task.Name, task.OnExit, dagCtx.boundaryID, dagCtx.tmplCtx)
-		if hasOnExitNode && (onExitNode == nil || !onExitNode.Completed() || err != nil) {
-			// The onExit node is either not complete or has errored out, return.
-			return
+	if node != nil && node.Fulfilled() {
+		if node.Completed() {
```
Comment (alexec) [inline workflow/controller/dag.go:381]: can you double-check this logic please? this has changed - which I think you intended - but please double check again

File: workflow/controller/steps.go [hunk 5]
```diff
@@ -255,11 +255,11 @@ func (woc *wfOperationCtx) executeStepGroup(stepGroup []wfv1.WorkflowStep, sgNod
 	for _, childNodeID := range node.Children {
 		childNode := woc.wf.Status.Nodes[childNodeID]
 		step := nodeSteps[childNode.Name]
-		if !childNode.Completed() {
+		if !childNode.Fulfilled() {
 			completed = false
-		} else {
+		} else if childNode.Completed() {
```
Comment (alexec) [inline workflow/controller/steps.go:260]: as above - this is change to logic - please double check you intended it

File: workflow/controller/dag.go (hunk 4 above)
Comment (simster7) [reply workflow/controller/dag.go:381]: This is intended, thanks for asking to make sure.

File: workflow/controller/steps.go (hunk 5 above)
Comment (simster7) [reply workflow/controller/steps.go:260]: This is intended, thanks for asking to make sure.

**Pull request merged**